File: src/main/java/idlezoo/game/domain/ZooBuildings.java
package idlezoo.game.domain;

import com.fasterxml.jackson.annotation.JsonPropertyOrder;
import org.immutables.value.Value;

@Value.Immutable
@JsonPropertyOrder(alphabetic = true)
public interface ZooBuildings {
    // Getters for JSON
    String getName();

    default double getIncome() {
        return getBuilding().income(getLevel()) * getNumber();
    }

    default double getNextCost() {
        return getBuilding().buildCost(getNumber());
    }

    default double getUpgradeCost() {
        return getBuilding().upgradeCost(getLevel());
    }

    int getLevel();

    int getNumber();

    int getLost();

    Building getBuilding();
}

File: src/main/kotlin/idlezoo/game/domain/ZooBuildings.kt
package idlezoo.game.domain;

import com.fasterxml.jackson.annotation.JsonPropertyOrder


@JsonPropertyOrder(alphabetic = true)
data class ZooBuildings(val level: Int, val number: Int, val lost: Int, val building: Building) {
    fun getIncome(): Double = building.income(level) * number
    fun getNextCost(): Double = building.buildCost(number)
    fun getUpgradeCost(): Double = building.upgradeCost(level)
    fun getName(): String = building.name
}